plugins/statuslines/lualine.nix

```diff
@@ -1,256 +1,401 @@
+              #   default = defaultName;
+              # };
+
+              icons_enabled = defaultNullOpts.mkBool true ''
+                Enables the display of icons alongside the component.
+              '';
+
+              icon = lib.nixvim.mkNullOrOption' {
+                type = either str (submodule {
+                  freeformType = attrsOf anything;
+
+                  # TODO: deprecate
+                  # options = {
+                  #   icon = mkOption {
+                  #     type = str;
+                  #     description = "Icon character.";
+                  #   };
+                  # };
+                });
+                description = "Defines the icon to be displayed in front of the component.";
+              };
+
+              separator = mkSeparatorsOption { };
+
+              color = defaultNullOpts.mkNullable (either attrs str) null ''
+                Defines a custom color for the component.
+
+                Default is the color defined by your theme for that section and mode.
+              '';
+
+              padding = defaultNullOpts.mkNullable (either int (submodule {
                 options = {
                   left = mkOption {
-                    type = types.int;
+                    type = int;
                     description = "left padding";
                   };
                   right = mkOption {
-                    type = types.int;
+                    type = int;
                     description = "left padding";
                   };
                 };
-              }
-            )) 1 "Adds padding to the left and right of components.";
-
-            fmt = helpers.mkNullOrLuaFn ''
-              A lua function to format the component string.
-
-              Example:
-              ```lua
-                function(text)
-                  return text .. "!!!"
-                end
-              ```
-            '';
-
-            extraConfig = mkOption {
-              type = types.attrs;
-              default = { };
-              description = "extra options for the component";
+              })) 1 "Adds padding to the left and right of components.";
+
+              fmt = lib.nixvim.mkNullOrLuaFn ''
+                A lua function to format the component string.
+
+                Example:
+                ```lua
+                  function(text)
+                    return text .. "!!!"
+                  end
+                ```
+              '';
+
+              # TODO: deprecate
+              # extraConfig = mkOption {
+              #   type = attrs;
+              #   default = { };
+              #   description = "extra options for the component";
+              # };
             };
-          };
-        })
-      )
-    ) "";
+          })
+        );
+      # TODO: best way to do this?
+      # Was setting up apply to handle the old way of transforming a config to the setup call
+      apply =
```

Comment:
@MattSturgeon Tried handling the deprecation for these nested options and not sure if it's required to be done this way or if you have a simpler way of handling this.. I'm also struggling to find a simple way to handle the deprecation messages unless i can just inline the warn or something here when we process the transformation ?

Reply:
I think it'd be easier to keep the old option and new option separate in this case. I.e. no rename module; instead just remove the `default` and copy over the value _if_ it is defined?

Perhaps the new option could also just be `attrsOf anything`, instead of being a freeform submodule type?

Reply:
Well, the copying over the defined value is what this apply is for. It'll do the transformation we used to do if they defined it using the old options. Unless, you meant there's an even easier way of copying those values over and transforming them without the apply.

We did a little bit of magic in this submodule with the `name`, `icon`, and `extraConfig` options to be flattened and unkeyed. I wanted to just show a message when I detected any of those usages. The hard part, for me, was that it's basically a submodule spread across a bunch of different options that i have to search for a proper deprecation warning on specific usages. But, if I could do the warning on the generic top level easily with a migration message about the new magic-less way of doing things that would be simpler..

Reply:
I meant that we need the apply here because we have a rename alias pointing to this option, right?

So by removing the rename alias, we could do the same thing the apply does but using a config definition instead.

Reply:
I'm thinking we can probably do this with `transitionType`, but note that coercedTo and transitionType don't support submodule-types on the lhs, so we'd have to use  something like:

```nix
let
  oldAttrs = [ "name" "icon" "extraConfig" ];
  isOldType = def: any (name: elem name oldAttrs) (attrNames def);
  oldType = (addCheck (attrsOf anything) isOldType) // {
    description = "attribute set containing ${concatStringsSep ", " oldAttrs}";
  };
  coerceFn = /* Essentially, do what you're currently doing in `apply` */;
in
listOf (
  transitionType
  (addCheck (attrsOf anything) isOldType)
  coerceFn
  (submodule {/* submodule options */})
```